step-functions/step-functions-client-api/src/main/java/step/functions/type/AbstractFunctionType.java

```diff
@@ -125,28 +132,58 @@ public T copyFunction(T function) throws FunctionTypeException {
-	 * @param props the map will be enriched with the propertyName id and version of the registered file that can be later used to retrieve the file
-	 * @param cleanable whether this version of the file can be cleaned-up at runtime
+	 * @param dynamicValue     the {@link DynamicValue} of the file's path to be registered
+	 * @param propertyName     the name of the property for which we register the file
+	 * @param props            the map will be enriched with the propertyName id and version of the registered file that can be later used to retrieve the file
+	 * @param cleanable        whether this version of the file can be cleaned-up at runtime
+	 * @param executionContext the current execution context (should be defined if the function is executing via ExecutionEngine with
+	 *                         execution-scope resource manager)
 	 * @throws RuntimeException
 	 */
-	protected void registerFile(DynamicValue<String> dynamicValue, String propertyName, Map<String, String> props, boolean cleanable) {
+	protected void registerFile(DynamicValue<String> dynamicValue, String propertyName, Map<String, String> props, boolean cleanable, AbstractStepContext executionContext) {
 		if(dynamicValue!=null) {
 			String filepath = dynamicValue.get();
 			if(filepath!=null && filepath.trim().length()>0) {
-				File file;
+				File file = null;
 				try {
-					// Using the file resolver cache here to avoid performance issues
-					// This method might be called at every function execution
-					file = fileResolverCache.get(filepath);
+					// in case of isolated execution, the execution context contains temporary in-memory resource manager
+					// we have to use this manager instead of the global one from fileResolver
+					if (executionContext != null) {
+						boolean resolvedFromExecutionContext = false;
+						ResourceManager executionContextResourceManager = getResourceManager(executionContext);
+						if (executionContextResourceManager == getResourceManager(null)) {
+							// if resource manager is the global one, it is better to use fileResolverCache for performance reason
+							file = fileResolverCache.get(filepath);
+						} else if (executionContextResourceManager instanceof LayeredResourceManager) {
+							// performance hack - if the resource manager is layered, but contains the only one resource manager
+							// and this resource manager equals to the global one, we can use cached file resolver
+							ResourceManager unwrappedManager = unwrapResourceManager((LayeredResourceManager) executionContextResourceManager);
+							if (unwrappedManager == getResourceManager(null)) {
+								file = fileResolverCache.get(filepath);
+							} else {
+								resolvedFromExecutionContext = true;
+								file = new FileResolver(executionContextResourceManager).resolve(filepath);
+							}
+						} else {
+							resolvedFromExecutionContext = true;
+							file = new FileResolver(executionContextResourceManager).resolve(filepath);

```

Comment:
@jeromecomte 

This sounds good. But for me it is not clear how to instantiate the `FileResolverCache` inside the `ExecutionContext`.
As I understood, in `AbstractFunctionType` we create an individual cache for each function type, and we apply the `FunctionTypeConfiguration` to `CacheBuilder` to build the cache properly.

In case we store the cache within the  `ExecutionContext` should we create some map `FunctionType` -> `FileResolverCache` in `ExecutionContext`?

Reply:
@jeromecomte 
Finally I've added the `fileResolverCache` field to the `AbstractStepContext` (configured with default concurrency level and maximum size).